File: switsol_ag/switsol_ag/report/personal_task_list/personal_task_list.py
# -*- coding: utf-8 -*-

# Copyright (c) 2017, Switsol AG and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe

from frappe import _
from frappe.utils import formatdate


def execute(filters=None):
    columns = get_column()
    data = prepare_data(filters)
    return columns, data


def get_column():
    return [
        _("Due Date") + ":Date:160",
        _("Expected Start Date") + ":Date:160",
        _("Responsible Person") + ":Link/User:240",
        _("Subject") + ":Data:120",
        _("Status") + ":Data:120",
        _("Name") + ":Data:120"
    ]


def get_data(filters):
    statuses = ["Open", "Waiting for Other Person", "Working"]
    if filters.get("done_tasks"):
        statuses.extend(["Done", "Closed", "Unnecessary"])
    responsible_person = ""
    if not filters.get("all_tasks"):
        responsible_person = frappe.session.user
    return frappe.db.sql("""
        SELECT
            exp_end_date,
            exp_start_date,
            _assign,
            name,
            status,
            subject,
            description
        FROM
            `tabTask`
        WHERE
            status IN (%s)
            AND ifnull(_assign,"") LIKE %s
        ORDER BY
            exp_end_date
        """ % (', '.join(['%s'] * len(statuses)), '%s'), tuple(statuses + ["%%%s%%" % responsible_person]), as_dict=1)


def prepare_data(filters):
    data = []
    tasks = get_data(filters)

    for t in tasks:
        assign = frappe.db.sql("""
            SELECT
                owner
            FROM 
                tabToDo
            WHERE reference_type='Task' AND reference_name=%s""", t.name, as_dict=1)
        assign = [i.owner for i in assign]
        row = [
            formatdate(t.exp_end_date),
            formatdate(t.exp_start_date),
            ', '.join(assign),
            str(t.subject) + "\n" + str(t.description),
            _(t.status),
            t.name
        ]
        data.append(row)

    return data

File: switsol_ag/switsol_ag/report/personal_task_list/personal_task_list.js
// Copyright (c) 2016, Switsol AG and contributors
// For license information, please see license.txt

frappe.query_reports["Personal Task List"] = {
	"filters": [
		{
			"fieldname": "all_tasks",
			"label": __("Alle Aufgaben (auch fremde) anzeigen"),
			"fieldtype": "Check",
			"default": 0,
			"reqd": 0
		},
		{
			"fieldname": "done_tasks",
			"label": __("Erledigte Aufgaben anzeigen"),
			"fieldtype": "Check",
			"default": 0,
			"reqd": 0
		},
	],
	formatter: function (row, cell, value, columnDef, dataContext, default_formatter) {
		value = default_formatter(row, cell, value, columnDef, dataContext);
		if(columnDef.id == __("Name")) {
			if(dataContext["Name"]) {
				value = "<a href=" + frappe.urllib.get_base_url() + "/desk#Form/Task/" + encodeURIComponent(dataContext["Name"])+" >" + dataContext["Name"] + "</a>";
			}
		}
		if(columnDef.id == __("Fälligkeitsdatum")) {
			if(dataContext["Name"]) {
				value = "<a href=" + frappe.urllib.get_base_url() + "/desk#Form/Task/" + encodeURIComponent(dataContext["Name"])+" >" + dataContext[__("Fälligkeitsdatum")] + "</a>";
			}
		}
		if(columnDef.id == __("Due Date")) {
			if(dataContext["Name"]) {
				value = "<a href=" + frappe.urllib.get_base_url() + "/desk#Form/Task/" + encodeURIComponent(dataContext["Name"])+" >" + dataContext[__("Due Date")] + "</a>";
			}
		}
		if(columnDef.id == __("Expected Start Date")) {
			if(dataContext["Name"]) {
				value = "<a href=" + frappe.urllib.get_base_url() + "/desk#Form/Task/" + encodeURIComponent(dataContext["Name"])+" >" + dataContext[__("Expected Start Date")] + "</a>";
			}
		}
		if(columnDef.id == __("Responsible Person")) {
			if(dataContext["Name"]) {
				value = "<a href=" + frappe.urllib.get_base_url() + "/desk#Form/Task/" + encodeURIComponent(dataContext["Name"])+" >" + dataContext[__("Responsible Person")] + "</a>";
			}
		}
		if(columnDef.id == __("Subject")) {
			if(dataContext["Name"]) {
				value = "<a href=" + frappe.urllib.get_base_url() + "/desk#Form/Task/" + encodeURIComponent(dataContext["Name"])+" >" + dataContext[__("Subject")] + "</a>";
			}
		}
		if(columnDef.id == __("Status")) {
			if(dataContext["Name"]) {
				value = "<a href=" + frappe.urllib.get_base_url() + "/desk#Form/Task/" + encodeURIComponent(dataContext["Name"])+" >" + dataContext[__("Status")] + "</a>";
			}
		}
		return value;
	},
}
